LXD daemon dies suddenly

<!--
Github issues are used for bug reports. For support questions, please use [our forum](https://discuss.linuxcontainers.org).

Please fill the template below as it will greatly help us track down your issue and reproduce it on our side.  
Feel free to remove anything which doesn't apply to you and add more information where it makes sense.
-->

# Required information

 * Distribution: Plamo Linux
 * Distribution version: 7.1
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.5.4
   * LXC version: 3.2.1
   * LXD version: 3.21
   * Storage backend in use: btrfs

# Issue description

LXD daemon dies unexpectedly. Nothing writes in the log when LXD is started nomally (with --debug).

I started with "--trace daemon", so I get the output:

```
panic: strings: negative Repeat count

goroutine 275 [running]:
strings.Repeat(0x1414c57, 0x1, 0xffffffffffffffff, 0xc0006ca028, 0x4)
        /usr/lib/go/src/strings/strings.go:533 +0x5aa
main.deviceNetlinkListener.func1.2(...)
        /tmp/build/src/github.com/lxc/lxd/lxd/devices.go:256
main.deviceNetlinkListener.func1(0x800, 0x18, 0xc000288660, 0xc0001c0120, 0xc0001c0180, 0xc0001c0240)
        /tmp/build/src/github.com/lxc/lxd/lxd/devices.go:266 +0x6fb
created by main.deviceNetlinkListener
        /tmp/build/src/github.com/lxc/lxd/lxd/devices.go:96 +0x210
```

I'm not sure what information you need, so please tell us what additional information you need.

--- a/lxd/device/unix_hotplug.go
+++ b/lxd/device/unix_hotplug.go
@@ -78,13 +78,13 @@ func (d *unixHotplug) Register() error {
 
 	// Handler for when a UnixHotplug event occurs.
 	f := func(e UnixHotplugEvent) (*deviceConfig.RunConfig, error) {
-		if !unixHotplugIsOurDevice(devConfig, &e) {
-			return nil, nil
-		}
-
 		runConf := deviceConfig.RunConfig{}
 
 		if e.Action == "add" {
+			if !unixHotplugIsOurDevice(devConfig, &e) {
+				return nil, nil
+			}
+
 			if e.Subsystem == "block" {
 				err := unixDeviceSetupBlockNum(state, devicesPath, "unix", deviceName, devConfig, e.Major, e.Minor, e.Path, false, &runConf)
 				if err != nil {

--- a/lxd/devices.go
+++ b/lxd/devices.go
@@ -227,17 +227,17 @@ func deviceNetlinkListener() (chan []string, chan []string, chan device.USBEvent
 
 			// unix hotplug device events rely on information added by udev
 			if udevEvent {
+				action := props["ACTION"]
+				if action != "add" && action != "remove" {
+					continue
+				}
+
 				subsystem, ok := props["SUBSYSTEM"]
 				if !ok {
 					continue
 				}
 
 				devname, ok := props["DEVNAME"]
-				if !ok {
-					continue
-				}
-
-				vendor, product, ok := ueventParseVendorProduct(props, subsystem, devname)
 				if !ok {
 					continue
 				}
@@ -252,18 +252,36 @@ func deviceNetlinkListener() (chan []string, chan []string, chan device.USBEvent
 					continue
 				}
 
+				vendor := ""
+				product := ""
+				if action == "add" {
+					vendor, product, ok = ueventParseVendorProduct(props, subsystem, devname)
+					if !ok {
+						continue
+					}
+				}
+
 				zeroPad := func(s string, l int) string {
 					return strings.Repeat("0", l-len(s)) + s
 				}
 
+				// zeropad
+				if len(vendor) < 4 {
+					vendor = zeroPad(vendor, 4)
+				}
+
+				if len(product) < 4 {
+					product = zeroPad(product, 4)
+				}
+
 				unix, err := device.UnixHotplugNewEvent(
-					props["ACTION"],
+					action,
 					/* udev doesn't zero pad these, while
 					 * everything else does, so let's zero pad them
 					 * for consistency
 					 */
-					zeroPad(vendor, 4),
-					zeroPad(product, 4),
+					vendor,
+					product,
 					major,
 					minor,
 					subsystem,
@@ -584,14 +602,14 @@ func getHidrawDevInfo(fd int) (string, string, error) {
 }
 
 func ueventParseVendorProduct(props map[string]string, subsystem string, devname string) (string, string, bool) {
+	vendor, vendorOk := props["ID_VENDOR_ID"]
+	product, productOk := props["ID_MODEL_ID"]
+
+	if vendorOk && productOk {
+		return vendor, product, true
+	}
+
 	if subsystem != "hidraw" {
-		vendor, vendorOk := props["ID_VENDOR_ID"]
-		product, productOk := props["ID_MODEL_ID"]
-
-		if vendorOk && productOk {
-			return vendor, product, true
-		}
-
 		return "", "", false
 	}
 
@@ -606,7 +624,7 @@ func ueventParseVendorProduct(props map[string]string, subsystem string, devname
 
 	defer file.Close()
 
-	vendor, product, err := getHidrawDevInfo(int(file.Fd()))
+	vendor, product, err = getHidrawDevInfo(int(file.Fd()))
 	if err != nil {
 		logger.Debugf("Failed to retrieve device info from hidraw device \"%s\"", devname)
 		return "", "", false

